### PR TITLE
chore(turbo-ignore): add console message of an unfriendly error

### LINKED ISSUE #1859: turbo-ignore: fails with an unfriendly error if the monorepo root `package.json` does not have the `name` field
### What version of Turborepo are you using?

1.4.5

### What package manager are you using / does the bug impact?

npm, pnpm, Yarn v1, Yarn v2/v3 (node_modules linker only)

### What operating system are you using?

Mac

### Describe the Bug

`turbo-ignore` fails with an unfriendly error if the monorepo root `package.json` does not have the `name` field.

I believe that in most projects it would probably never be a problem, but my team didn't put the `name` field in the root and tried to debug it by poking `turbo-ignore` a lot, but it just didn't work, so I figured it out by going to the source.

The reason for this is that the `scope` field returned by `getScopeFromPath` in `turbo-utils`, which `turbo-ignore` uses, depends on the `name` field in `package.json`.

https://github.com/vercel/turborepo/blob/04b077a1582badd4c8378b8357691b5795dad7ee/packages/turbo-ignore/src/index.ts#L23

If the `scope` of `argsScope` is not found, it will fallback to `pathScope`, but unfortunately in this situation the `scope` of `pathScope` is also `null`. The possible error in this case is almost always the missing `name` field, and in my opinion just a polite error message is fine.

https://github.com/vercel/turborepo/blob/1116799546f688ec17cd9fe708dae4413b182955/packages/turbo-utils/src/getScopeFromPath.ts#L6-L13

I think it makes sense to just make sure that `argsScope.scope` and `scope` aren't found here, and then give a hint for users.

https://github.com/vercel/turborepo/blob/04b077a1582badd4c8378b8357691b5795dad7ee/packages/turbo-ignore/src/index.ts#L24-L30

### Expected Behavior

It needs printing a more friendly error or different utility implementations.

### To Reproduce

Remove the `name` field from `package.json` in the root of your current turborepo monorepo.

### PR DESCRIPTION
**Pull request by t-i-0414**
## issue

- Fixes #1859 

## description

fix above issue by adding console message.

### CODE REVIEW

File: packages/turbo-ignore/src/index.ts [hunk 1]
```diff
@@ -25,6 +25,11 @@ if (!scope) {
   console.error(
     "Error: app scope not found. turbo-ignore inferencing failed, proceeding with build."
   );
+  if (!pathScope.scope) {
+    console.error(
+      "Hint: turbo-ignore cannot process because the monorepo root package.json does not have the name field.\nset or pass it as an explicit argument to turbo-ignore"
```
Comment (mehulkar) [inline packages/turbo-ignore/src/index.ts:30]:
```suggestion
      "Error: the root package.json is missing the \"name\" field.\nSet this field or pass "scope" as an explicit argument to turbo-ignore."
```

side question: what exactly would be passed as an explicit arg?

**Review by mehulkar — COMMENTED** [review]
I haven't used `turbo-ignore` much yet, so I'd want @tknickman to double check the error message too

File: packages/turbo-ignore/src/index.ts (hunk 1 above)
Comment (t-i-0414) [reply packages/turbo-ignore/src/index.ts]: @mehulkar 
thank you for your suggestion. I commit your suggestion code and refactor it a little. 

5b14d27
6da7a20
edddb8c

> side question: what exactly would be passed as an explicit arg?

I read in the code that turbo-ignore is supposed to process with the value of the name field in package.json as an arg.
Finally, this argument is supposed to be passed to `turbo` command  as a filter arg.

So I think the following use case might be possible.

### 1. package.json has the name field ("name": "hoge")

```bash
cd ./packages/hoge
npx turbo-ignore
≫ Using Turborepo to determine if this project is affected by the commit...
≫ Inferred `hoge` as scope from "/Users/t-i-0414/project/packages/hoge/package.json"
≫ Analyzing results of `npx turbo run build --filter=hoge...[HEAD^] --dry=json`...
≫ This project and its dependencies are not affected
≫ Ignoring the change
```

### 2. package.json does not have the name field

```bash
cd ./packages/hoge
npx turbo-ignore  
≫ Using Turborepo to determine if this project is affected by the commit...
Error: app scope not found. turbo-ignore inferencing failed, proceeding with build.
```

### 3. package.json does not have the name field but passing an arg

```bash
cd ./packages/hoge
npx turbo-ignore hoge
≫ Using Turborepo to determine if this project is affected by the commit...
≫ Inferred `hoge` as scope from arguments
≫ Analyzing results of `npx turbo run build --filter=hoge...[HEAD^] --dry=json`...
≫ This project and its dependencies are not affected
≫ Ignoring the change
```

### 4. package.json has the name field (name: hoge) and passing an arg fuga

```bash
cd ./packages/hoge
npx turbo-ignore fuga
≫ Using Turborepo to determine if this project is affected by the commit...
≫ Inferred `fuga` as scope from arguments
≫ Analyzing results of `npx turbo run build --filter=fuga...[HEAD^] --dry=json`...
≫ This project and its dependencies are not affected
≫ Ignoring the change
```

I assume that this PR will take care of case 2 as following.
↓
```
cd ./packages/hoge
npx turbo-ignore
≫ Using Turborepo to determine if this project is affected by the commit...
Error: app scope not found. turbo-ignore inferencing failed, proceeding with build.
Error: the package.json is missing the "name" field.
Set this field or pass "scope" as an explicit argument to turbo-ignore.
```

File: packages/turbo-ignore/src/index.ts [hunk 2]
```diff
@@ -25,6 +25,11 @@ if (!scope) {
   console.error(
     "Error: app scope not found. turbo-ignore inferencing failed, proceeding with build."
   );
+  if (!pathScope.scope) {
+    console.error(
+      'Error: the package.json is missing the "name" field.\nSet this field or pass "scope" as an explicit argument to turbo-ignore.'
```
Comment (mehulkar) [inline packages/turbo-ignore/src/index.ts:30]:
```suggestion
      'Error: the package.json is missing the "name" field.\nSet this field or pass the --scope flag to turbo-ignore.'
```

**Review by mehulkar — APPROVED** [review]

File: packages/turbo-ignore/src/index.ts [hunk 3]
```diff
@@ -25,6 +25,11 @@ if (!scope) {
   console.error(
     "Error: app scope not found. turbo-ignore inferencing failed, proceeding with build."
   );
+  if (!pathScope.scope) {
+    console.error(
+      'Error: the package.json is missing the "name" field.\nSet this field or pass the --scope flag to turbo-ignore.'
```
Comment (smorimoto) [inline packages/turbo-ignore/src/index.ts:30]: Note that the `--scope` flag is not available. `turbo-ignore` just takes the first argument...

**Pull request merged**